core/classes/User.php

```diff
@@ -404,127 +404,56 @@ public function getSignature() {
-
-                        break;
-
-                    case 'mc-heads':
-                        if ($perspective == 'face')
-                            return 'https://mc-heads.net/avatar/' . Output::getClean($uuid) . '/' . $size;
-                        else
-                            return 'https://mc-heads.net/head/' . Output::getClean($uuid) . '/' . $size;
-
-                        break;
-
-                    case 'minotar':
-                        if ($perspective == 'face')
-                            return 'https://minotar.net/helm/' .  Output::getClean($uuid) . '/' . $size . '.png';
-                        else
-                            return 'https://minotar.net/cube/' .  Output::getClean($uuid) . '/' . $size . '.png';
-
-                        break;
-
-                    case 'visage':
-                        if ($perspective == 'face')
-                            return 'https://visage.surgeplay.com/face/' . $size . '/' . Output::getClean($uuid);
-                        else if ($perspective == 'bust')
-                            return 'https://visage.surgeplay.com/bust/' . $size . '/' . Output::getClean($uuid);
-                        else
-                            return 'https://visage.surgeplay.com/head/' . $size . '/' . Output::getClean($uuid);
-
-                        break;
+            if (file_exists(ROOT_PATH . '/uploads/avatars/defaults/' . DEFAULT_AVATAR_IMAGE)) {
+                return ($full ? rtrim(Util::getSelfURL(), '/') : '') . ((defined('CONFIG_PATH')) ? CONFIG_PATH . '/' : '/') . 'uploads/avatars/defaults/' . DEFAULT_AVATAR_IMAGE;
+            }
+        }
 
-                    case 'cravatar':
-                    default:
-                        if ($perspective == 'face')
-                            return 'https://cravatar.eu/helmavatar/' . Output::getClean($uuid) . '/' . $size . '.png';
-                        else
-                            return 'https://cravatar.eu/helmhead/' . Output::getClean($uuid) . '/' . $size . '.png';
-                        break;
-                }
-            } else {
-                // Fall back to cravatar
-                return 'https://cravatar.eu/helmavatar/' . Output::getClean($uuid) . '/' . $size . '.png';
+        // If all else fails, or custom avatars are disabled or default avatar type is 'minecraft', get their MC avatar
+        if ($this->data()->uuid != null && $this->data()->uuid != 'none') {
+            $uuid = $this->data()->uuid;
+        } else {
+            $uuid = $this->data()->username;
+            // Fallback to steve avatar if they have an invalid uuid
+            if (preg_match("#[^][_A-Za-z0-9]#", $uuid)) {
+                $uuid = 'Steve';
             }
         }
+
+        return AvatarSource::getAvatarFromUUID($this->data()->uuid, $size);
```

Comment:
Should the first parameter be `$uuid` instead of pulling it from data?

Reply:
good catch